app/controllers/admin/deploy_groups_controller.rb

```diff
@@ -82,50 +82,88 @@ def create_all_stages_preview
   # No more than one stage, per project, per deploy_group
   # Note: you can call this multiple times, and it will create missing stages, but no redundant stages.
   def create_all_stages
-    _, missing_stages = stages_for_creation
-    missing_stages.each do |template_stage|
-      create_stage_with_group(template_stage)
+    self.class.create_all_stages(deploy_group)
+
+    redirect_to [:admin, deploy_group]
+  end
+
+  def merge_all_stages
+    preexisting_stages, = stages_for_creation
+    template_stages = deploy_group.environment.template_stages.all
+
+    preexisting_stages.each do |stage|
+      template_stage = template_stages.detect { |ts| ts.project_id == stage.project.id }
+      merge_stage(stage, template_stage)
     end
 
     redirect_to [:admin, deploy_group]
   end
 
+  def self.create_all_stages(deploy_group)
+    _, missing_stages = stages_for_creation(deploy_group)
+    missing_stages.each do |template_stage|
+      create_stage_with_group(template_stage, deploy_group)
+    end
+  end
+
   private
 
-  # returns a list of stages already created and list of stages to create (through their template stages)
+  def merge_stage(stage, template_stage)
+    return unless template_stage
+    return if template_stage.deploy_groups.include?(stage)
+    return if stage.is_template
+    return if stage.deploy_groups.count > 1
+
+    template_stage.deploy_groups += stage.deploy_groups
+    template_stage.next_stage_ids.delete(stage.id)
+    template_stage.save!
+
+    # these don't respect soft_deletion (yet).
+    DeployGroupsStage.where(stage: stage).delete_all
```

Comment:
@grosser I just delete these as we won't need them anymore. The alternative is to add them to soft_delete, which will require, at the least, a database migration. What do you think?